Services

CrazyTapok opened the pull request:
1. Screenshot: 
![image](https://github.com/user-attachments/assets/d2ba12b0-2a51-4be2-a04a-abc2e5eaa589)


2. What has been implemented:
  - [x] Add generic service 
  - [x] Add service for each entity 
  - [x] Add interface for services 
  - [x] Fix the links between entities 
  - [x] Add IsDeleted propery in models 
  - [x] Change method DeleteAsync in Repository
  - [x] Change method DeleteEntity in RepositoryTest
  - [x] Change configuration 
  - [x] Create new migration 
  - [x] Add repositories for each entity 
  - [x] Add new tests.

CrazyTapok committed:
feat: add IsDeleted propery in models
Add IsDeleted propery in models, create new migration, change method DeleteAsync in Repository, change method DeleteEntity in RepositoryTest, change entity configuragtion

CrazyTapok committed:
feat: add services

Add generic service, add service for each entity, add interface for services, fix the links between entities, change configuration, create new migration, add repositories for each entity, add new tests.

svirzhevskiy commented on src/TaskManager.API/TaskManager.API.csproj:
```diff
@@ -7,7 +7,7 @@
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="8.0.8">
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="9.0.0">
```
пошла жара?) Может тогда вообще на .net 9 обновить?)

CrazyTapok replied:
+

svirzhevskiy commented on src/TaskManager.Core/Infrastructure/ServiceCollection.cs:
```diff
@@ -0,0 +1,18 @@
+        services.AddTransient<IEmployeeService, EmployeeService>(); 
+        services.AddTransient<ICompanyService, CompanyService>();
+        services.AddTransient<IProjectService, ProjectService>();
+        services.AddTransient <ITaskService, TaskService>();
```
```suggestion
        services.AddTransient<ITaskService, TaskService>();
```

svirzhevskiy commented on src/TaskManager.Core/Interfaces/Data/IProjectRepository.cs:
```diff
@@ -0,0 +1,9 @@
+
+public interface IProjectRepository : IRepository<Project> 
+{
+    Task<IEnumerable<Project>> FindAsync(Expression<Func<Project, bool>> predicate, CancellationToken cancellationToken);
```
почему везде одинаковый FindAsync? его же можно в дженерик переместить

svirzhevskiy commented on src/TaskManager.Core/Interfaces/Data/IRepository.cs:
```diff
@@ -0,0 +1,10 @@
+
+public interface IRepository<TModel> where TModel : class
+{
+    Task<List<TModel>> GetAllAsync(CancellationToken cancellationToken);
```
CancellationToken часто делают необязательным параметром, но можно и так

svirzhevskiy commented on src/TaskManager.Core/Interfaces/Services/ICompanyService.cs:
```diff
@@ -0,0 +1,7 @@
+
+namespace TaskManager.Core.Interfaces.Services;
+
+internal interface ICompanyService : IService<Company>
```
вроде же не обязательно создавать пустой интерфейс. Просто пользуешься дженериком везде

svirzhevskiy commented on src/TaskManager.Core/Models/Company.cs:
```diff
@@ -9,4 +9,6 @@ public class Company
 
     public List<Employee> Employees { get; set; }
+
+    public bool IsDeleted { get; set; }
```
бля, вот это конечно не оч красиво выходит. Такое проперти хочется держать только в модели для базы и не видеть в моделях кора. Номы этим жертвуем :(

svirzhevskiy commented on src/TaskManager.Core/Services/CompanyService.cs:
```diff
@@ -0,0 +1,47 @@
+
+namespace TaskManager.Core.Services;
+
+public class CompanyService : ICompanyService
```
делай базовую реализацию дженерика с общими методами и наследуйся от нее в сервисах для конкретной модели. Много дублирования кода уберешь. У тебя все эти сервисы почти идентичны, за исключением модели. Дженерик же может быть не только интерфейс, но и реализация

svirzhevskiy commented on src/TaskManager.Core/Services/CompanyService.cs:
```diff
@@ -0,0 +1,47 @@
+
+    public async Task<List<Company>> ListAllAsync(CancellationToken cancellationToken) 
+    { 
+        return await _companyRepository.GetAllAsync(cancellationToken); 
```
await тут не обязателен, можно просто таску возвращать

svirzhevskiy commented on src/TaskManager.Core/Services/CompanyService.cs:
```diff
@@ -0,0 +1,47 @@
+        var entity = await _companyRepository.GetByIdAsync(id, cancellationToken); 
+       
+        if (entity == null) 
+            throw new ValidationException($"Entity with id {id} not found", ""); 
```
это не ValidationException а какой-нибудь not found

Commit:
refactor: fix remarks after the review

The project has been translated to version 9.0. Removed repository interfaces for each entity, moved the FindAsync method to the generic interface. The CancellationToken parameter is made optional in all methods where it is used. Removed ICompanyService. A generic Service class has been created. The general functionality has been moved to the Service class from entity services. Class ValidationException renamed NotFoundExeption. Tests have been changed.

CrazyTapok committed:
docs: update gitignore

svirzhevskiy commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -1,50 +1,63 @@
+    public async Task<IEnumerable<TModel>> FindAsync(Expression<Func<TModel, bool>> predicate, CancellationToken? cancellationToken)
+    {
+        return await _dbSet.Where(predicate).ToListAsync(cancellationToken ?? CancellationToken.None);
+    }
```
```suggestion
    public Task<List<TModel>> FindAsync(Expression<Func<TModel, bool>> predicate, CancellationToken cancellationToken = default)
    {
        return _dbSet.Where(predicate).ToListAsync(cancellationToken);
    }
```

Я это имелл ввиду, когда писал про ctoken :) Налабл его делать не нужно. Вообще стремимся избегать налов в коде

svirzhevskiy replied:
и все таки почему ты не хочешь использовать CancellationToken cancellationToken = default?🤔
Это, конечно, дело вкуса, но дает возможность не использовать его там, где его нет и выглядит красивей

CrazyTapok replied:
я просто не знал про такое

svirzhevskiy commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -1,50 +1,63 @@
         {
-            _dbSet.Remove(entity);
-            
+            typeof(TModel).GetProperty("IsDeleted")?.SetValue(entity, true);
```
рефлексия бьет по перформансу всегда, старайся ее избегать.
Чтобы решить вопрос здесь, нужно создать базовый класс, который будет содержать общие свойства для всех моделей и использовать его в дженериках

svirzhevskiy commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -1,50 +1,63 @@
+    public Task<TModel> GetByIdAsync(Guid id, CancellationToken? cancellationToken)
     {
-        return _dbSet.FindAsync(id).AsTask();
+        return _dbSet.FindAsync(id, cancellationToken ?? CancellationToken.None).AsTask();
```
почему здесь избегаешь ValueTask?
Почитай чем отличается от обычной и заиспользуй

svirzhevskiy commented on src/TaskManager.Infrastructure/Repositories/TaskRepository.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace TaskManager.Infrastructure.Repositories;
+
+internal class TaskRepository : Repository<Task>
```
Ответ на твой вопрос по FindAsync: Если есть специфик серфис, то и интерфейс для него свой фигач, иначе ты реально их не отличишь потом в DI

svirzhevskiy replied:
при этом в сервисах ты почему-то это сделал(～￣▽￣)～

svirzhevskiy commented on src/TaskManager.Infrastructure/Repositories/ProjectRepository.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace TaskManager.Infrastructure.Repositories;
+
+internal class ProjectRepository : Repository<Project>
```
В каком-то там дотнете добавили primary constructor. Мне нравится использовать их в небольших классах
https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/primary-constructors#dependency-injection

CrazyTapok replied:
это можно применить ко всем классам сервисов и репозиториям?

svirzhevskiy commented on src/TaskManager.Infrastructure/Repositories/ProjectRepository.cs:
```diff
@@ -0,0 +1,19 @@
+    {
+    }
+
+    public async Task<IEnumerable<Project>> FindAsync(Expression<Func<Project, bool>> predicate, CancellationToken? cancellationToken)
```
Ты везде IEnumerable как возвращаемое значение поставил, чтобы потом все равно кастить в лист? 😁
![image](https://github.com/user-attachments/assets/3413fafb-be45-45fa-82bd-9f1f917eaec6)
Поставь лист и не парься. И сможешь возвращать таску, а не эвейтить тут

svirzhevskiy commented on src/TaskManager.Core/Infrastructure/ServiceCollection.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace TaskManager.Core.Infrastructure;
+
+public static class ServiceCollection
```
```suggestion
public static class ServiceCollectionExtensions
```

svirzhevskiy commented on src/TaskManager.Core/Infrastructure/ServiceCollection.cs:
```diff
@@ -0,0 +1,19 @@
+        services.AddTransient<IEmployeeService, EmployeeService>(); 
+        services.AddTransient<IProjectService, ProjectService>();
+        services.AddTransient<ITaskService, TaskService>();
+        services.AddTransient(typeof(IService<Company>), typeof(Service<Company>)); 
```
```suggestion
        services.AddTransient<IService<Company>, Service<Company>>(); 
```

svirzhevskiy commented on src/TaskManager.Core/Infrastructure/ServiceCollection.cs:
```diff
@@ -0,0 +1,19 @@
+    {
+        services.AddTransient<IEmployeeService, EmployeeService>(); 
+        services.AddTransient<IProjectService, ProjectService>();
+        services.AddTransient<ITaskService, TaskService>();
```
почему транзиент? я бы скоупд по дефолту делал

CrazyTapok committed:
refactor: fix remarks after the review

The gitignore file has been changed. The primary constructor is applied in the classes of repositories and services. Renamed Service Collection to Service Collection Extensions. The dependency lifecycle has been changed. Fixed dependencies.  Added a basic model for entities. Migration has been updated.

CrazyTapok committed:
refactor: fix remarks after the review
Replaced primary constructor to regular

svirzhevskiy commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -1,50 +1,60 @@
-
-    public Task<TModel> GetByIdAsync(Guid id)
+     
+    public ValueTask<TModel> GetByIdAsync(Guid id, CancellationToken cancellationToken)
```
```suggestion
    public ValueTask<TModel?> GetByIdAsync(Guid id, CancellationToken cancellationToken)
```

svirzhevskiy commented on src/TaskManager.Infrastructure/Repositories/Repository.cs:
```diff
@@ -1,50 +1,60 @@
+            entity.IsDeleted = true;
+
+            context.Entry(entity).State = EntityState.Modified; 
+            await context.SaveChangesAsync(); 
```
```suggestion
            await context.SaveChangesAsync(cancellationToken); 
```

svirzhevskiy commented on src/TaskManager.Core/Services/Service.cs:
```diff
@@ -0,0 +1,38 @@
+        var entity = await repository.GetByIdAsync(id, cancellationToken);
+
+        if (entity == null)
+            throw new NotFoundException($"Entity with id {id} not found", "");
```
не вкуриваю кастомный эксепшн. Ты сделал свой NotFoundException со свойством Property, которое заполняешь пустой строкой?

CrazyTapok replied:
Да, пока не знаю что туда передавать. Пока просто заложил

svirzhevskiy replied:
то есть ты это заложил сам не знаешь для чего? мб удалить тогда?

CrazyTapok replied:
Удалить property или сам класс?

svirzhevskiy commented on src/TaskManager.Core/Services/EmployeeService.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace TaskManager.Core.Services;
+
+public class EmployeeService : Service<Employee>, IEmployeeService 
```
для чего имплементации сервисов публичные? не выходит сделать интернал?

CrazyTapok replied:
выходит

svirzhevskiy commented on src/TaskManager.Core/Services/EmployeeService.cs:
```diff
@@ -0,0 +1,26 @@
+
+    public Task<List<Employee>> GetEmployeesByProjectIdAsync(Guid projectId, CancellationToken cancellationToken) 
+    {
+        return _employeeRepository.FindAsync(t => t.Projects.Any(t => t.Id == projectId), cancellationToken); 
```
косметический коммент чисто от меня: мне не нравится, когда в лямбдах позволяют себе называть переменные отпизды🥲 Почему переменные не в люмбдах нормально называем, а в них можно просто букву поставить? Могу еще принять, когда в цепочке только одна лямюда и буква это эта хотя бы первая из реального названия, но все равно хуйня, по-моему. 
Я сам так жил, но теперь против этого.
Имхо не сложно написать целое слово. Что здесь значит t? Почему t? Почему две разных переменных, одна означает employee, а другая project и обе называются t, что перекрывает друг друга?
```suggestion
        return _employeeRepository.FindAsync(employee => employee.Projects.Any(project => project.Id == projectId), cancellationToken); 
```

svirzhevskiy commented on src/TaskManager.Core/Services/EmployeeService.cs:
```diff
@@ -0,0 +1,26 @@
+    public Task<List<Employee>> GetEmployeesByProjectIdAsync(Guid projectId, CancellationToken cancellationToken) 
+    {
+        return _employeeRepository.FindAsync(t => t.Projects.Any(t => t.Id == projectId), cancellationToken); 
+        
```
```suggestion
```

CrazyTapok replied:
случайный коммент?

svirzhevskiy replied:
пустая строка лишняя. Не случайный

svirzhevskiy commented on src/TaskManager.Core/Models/Task.cs:
```diff
@@ -18,8 +16,8 @@ public class Task
     public Guid CreateEmployeeId { get; set; }
```
С точки зрения англа, конечно, лучше CreatedBy и employee можно убрать, если у тебя все эмплои :)
```suggestion
    public Guid CreatedById { get; set; }
```

svirzhevskiy replied:
необязательный коммент*

svirzhevskiy commented on src/TaskManager.Infrastructure/EF/DBContext.cs:
```diff
@@ -12,18 +12,21 @@ internal class DBContext : DbContext
-        Database.EnsureDeleted();
-        Database.EnsureCreated();
+        //Database.EnsureDeleted();
+        //Database.EnsureCreated();
```
можно это как-то поправить? Чтобы оно выполняло скрипты, только когда нужно. Например, когда базы нет

CrazyTapok replied:
думаю можно

svirzhevskiy commented:
заглянул в тесты, а ты их тоже от пизды херачишь, не годиться

svirzhevskiy commented on src/TaskManager.Tests/Core/Services/EmployeeServiceTests.cs:
```diff
@@ -0,0 +1,78 @@
+
+            var employees = new List<Employee>
+            {
+                new Employee { Id = Guid.NewGuid(), Name = "Employee1", Projects = new List<Project> { new Project { Id = projectId } } },
```
```suggestion
                _fixture.Build<Employee>().With(employee => employee.Projects, [_fixture.Build<Project>().With(project => project.Id, projectId).Create()]).Create(),
```

svirzhevskiy commented on src/TaskManager.Tests/Core/Services/EmployeeServiceTests.cs:
```diff
@@ -0,0 +1,78 @@
+                new Employee { Id = Guid.NewGuid(), Name = "Employee3", Projects = new List<Project> { new Project { Id = Guid.NewGuid() } } }
+            };
+
+            _mockRepo.Setup(repo => repo.FindAsync(It.IsAny<Expression<Func<Employee, bool>>>(), _cancellationToken))
+                .ReturnsAsync((Expression<Func<Employee, bool>> predicate, CancellationToken token) =>
+                {
+                    return employees.Where(predicate.Compile()).ToList();
+                });
```
```suggestion
            _mockRepo.Setup(repo => repo.FindAsync(It.IsAny<Expression<Func<Employee, bool>>>(), _cancellationToken))
                .ReturnsAsync((Expression<Func<Employee, bool>> predicate, CancellationToken _) =>
                    employees.Where(predicate.Compile()).ToList());
```

svirzhevskiy commented on src/TaskManager.Tests/Core/Services/EmployeeServiceTests.cs:
```diff
@@ -0,0 +1,78 @@
+        [Fact] 
+        public async Task GetEmployeesByCompanyIdAsync() 
+        { 
+            var companyId = Guid.NewGuid(); 
+            var employees = new List<Employee> 
+            { 
+                new Employee { Id = Guid.NewGuid(), Name = "Employee1", CompanyId = companyId }, 
+                new Employee { Id = Guid.NewGuid(), Name = "Employee2", CompanyId = companyId }, 
+                new Employee { Id = Guid.NewGuid(), Name = "Employee3", CompanyId = Guid.NewGuid() } 
+            }; 
+            
+            _mockRepo.Setup(repo => repo.FindAsync(It.IsAny<Expression<Func<Employee, bool>>>(), _cancellationToken))
+                .ReturnsAsync((Expression<Func<Employee, bool>> predicate, CancellationToken token) => 
+                { 
+                    return employees.Where(predicate.Compile()).ToList(); 
+                }); 
+            
+            var result = await _service.GetEmployeesByCompanyIdAsync(companyId, _cancellationToken); 
+            
+            Assert.Equal(2, result.Count); 
+            Assert.All(result, employee => Assert.Contains(employee, employees)); 
```
```suggestion
           //Arrange
            var companyId = Guid.NewGuid(); 
            var employees = new List<Employee> 
            { 
                new Employee { Id = Guid.NewGuid(), Name = "Employee1", CompanyId = companyId }, 
                new Employee { Id = Guid.NewGuid(), Name = "Employee2", CompanyId = companyId }, 
                new Employee { Id = Guid.NewGuid(), Name = "Employee3", CompanyId = Guid.NewGuid() } 
            }; 
            
            _mockRepo.Setup(repo => repo.FindAsync(It.IsAny<Expression<Func<Employee, bool>>>(), _cancellationToken))
                .ReturnsAsync((Expression<Func<Employee, bool>> predicate, CancellationToken token) => 
                { 
                    return employees.Where(predicate.Compile()).ToList(); 
                }); 
            
            //Act
            var result = await _service.GetEmployeesByCompanyIdAsync(companyId, _cancellationToken); 
            
            //Assert
            Assert.Equal(2, result.Count); 
            Assert.All(result, employee => Assert.Contains(employee, employees)); 
```
так в тестах для красоты пишут

svirzhevskiy commented on src/TaskManager.Tests/Core/Services/ServiceTests.cs:
```diff
@@ -0,0 +1,95 @@
+public class ServiceTests
+{
+    private readonly Mock<IRepository<TestModel>> _repositoryMock;
+    private readonly IService<TestModel> _service;
```
```suggestion
    private readonly Service<TestModel> _service;
```
тестим реализацию

svirzhevskiy commented on src/TaskManager.Tests/Infrastructure/EF/TestDBContext.cs:
```diff
@@ -1,11 +1,30 @@
+            modelBuilder.ApplyConfigurationsFromAssembly(assembly); 
+        } 
+        
+        base.OnModelCreating(modelBuilder); }
```
```suggestion
        base.OnModelCreating(modelBuilder); 
   }
```

svirzhevskiy commented on src/TaskManager.Tests/Infrastructure/Repositories/EmployeeRepositoryTests.cs:
```diff
@@ -0,0 +1,50 @@
+
+        var result = await _employeeRepository.FindAsync(predicate, _cancellationToken);
+
+        Assert.Equal(matchingEmployees.Count, result.Count());
```
```suggestion
        Assert.Equal(matchingEmployees.Count, result.Count);
```

svirzhevskiy commented on src/TaskManager.Tests/Infrastructure/Repositories/EmployeeRepositoryTests.cs:
```diff
@@ -0,0 +1,50 @@
+        await _context.Set<Employee>().AddRangeAsync(employees);
+        await _context.SaveChangesAsync();
+
+        Expression<Func<Employee, bool>> predicate = t => t.Name.Contains("BestEmployee");
```
"BestEmployee" выносим в константу, тк используешь потом для проверки

svirzhevskiy commented on src/TaskManager.Tests/Infrastructure/Repositories/RepositoryTests.cs:
```diff
@@ -1,22 +1,24 @@
-    private readonly DbContextOptions<TestDBContext> _options;
+    private readonly DbContextOptions<DBContext> _options;
     private readonly TestDBContext _context;
     private readonly IRepository<TestModel> _repository;
```
```suggestion
    private readonly Repository<TestModel> _repository;
```

svirzhevskiy commented:
на каждой мелочи останавливаемся, но это в воспитательных целях😊

svirzhevskiy commented:
сообщения коммитов, кстати, нравятся

CrazyTapok committed:
refactor: fix remarks after the review
Add using cancellationToken = default. Сhanged the access modifier for the implementation of services to internal. Variables in lambda expressions have been renamed. Added checking for the existence of a database for the application before creating. Fixed test issues.

CrazyTapok committed:
refactor: fix remarks after the review
An empty line has been removed. Removed property from class NotFoundException.
